### ISSUE TITLE
Store brightness to an EFI variable and let DRM core discover it

### ISSUE THREAD

**Issue by superm1** [systemd/systemd#34574]
### Is your feature request related to a problem? Please describe

Systemd provides a service that saves the brightness value to the filesystem and a unit that is run on bootup to restore it.  DRM drivers however also have their own default values that they set brightness to during boot up which can be a different value.  This leads to brightness coming up at one level and changing during the boot process.

### Describe the solution you'd like

At least for EFI systems I was thinking that using the efivarfs could be a nice way to avoid this.  When systemd saves the brightness level, it can _also_ be stored to an EFI variable at shutdown instead of the filesystem.

DRM core can grow a helper that can parse this EFI variable and that DRM drivers can call this to determine the initial brightness to program instead of a hardcoded default from the driver.

### The systemd version you checked that didn't have the feature you are asking for

256

**Comment by poettering** [comment]
We generally try to make EFI var changes only on install/update, but not during regular user interaction, or reboots or such, because supposedly EFI var backing is kinda shit on various devices.

it appears to me the DRM drivers should not change the brightness, and continue to use what the firmware sets, until systemd-backlight then restores things once /var/ is available.

Also, by storing things in an EFI var instead of in a file in /var/ we'll only making the window shorter (i.e. it will be a shorter time when a DRM device pops up, and we set the brightness), we are not going to be able to remove the window altogether (which would require a mechanism for providing input to DRM drivers for this, so that it never sets anything other than the target brightness).

Which hw is this about? Have you contacted the driver maintainers/DRM maintainers about this, and asked them what their take on this is?

**Comment by superm1** [comment]
> We generally try to make EFI var changes only on install/update, but not during regular user interaction, or reboots or such, because supposedly EFI var backing is kinda shit on various devices.

Even on modern stuff?  I wonder if drawing a line in the sand and say 2022 onwards or something is OK?

> it appears to me the DRM drivers should not change the brightness, and continue to use what the firmware sets, until systemd-backlight then restores things once /var/ is available.

Right; I tried to do that already:
https://github.com/torvalds/linux/commit/2fe87f54abdc5c29bd2ebc301964926ea27b3d89

My _hope_ is that if efivarfs stored brightness values then the moment plymouth comes up (when the driver loads) it can go right to the right brightness.

> Also, by storing things in an EFI var instead of in a file in /var/ we'll only making the window shorter (i.e. it will be a shorter time when a DRM device pops up, and we set the brightness), we are not going to be able to remove the window altogether (which would require a mechanism for providing input to DRM drivers for this, so that it never sets anything other than the target brightness).

Right; this is exactly my suggestion.  If we can do the systemd side of it to store it to efivarfs then DRM core can look that up and drivers can start out with the right value _immediately_.

> Which hw is this about? Have you contacted the driver maintainers/DRM maintainers about this, and asked them what their take on this is?

My observations and thoughts on this were from a Framework 13 AMD.  I've not discussed it on dri-devel, I figured I would start here.  If you're open to it then maybe we can do an RFC discussion on dri-devel.

**Comment by bluca** [comment]
> Even on modern stuff? I wonder if drawing a line in the sand and say 2022 onwards or something is OK?

Eh, we keep hearing horror stories about firmware implementations on cheap devices...

As long as we can do this from the initrd that should be good enough, no? We could save the value in a credential in the ESP so that plymouth can load it very early at boot. There might be a second or two where it's mismatched, but on balance it seems worth to avoid risks, especially as brigthness is something that is likely to be updated often.

**Comment by superm1** [comment]
> Eh, we keep hearing horror stories about firmware implementations on cheap devices...

But on _modern_ cheap devices?  Everyone uses the same 4-5 IBVs.  Modern stuff will be on a modern stack.

> We could save the value in a credential in the ESP so that plymouth can load it very early at boot.

Since the distro can choose to compile the driver into the kernel, put it in the initrd or put it in the rootfs there is no guarantee on the order of events.  I would argue if it needs to come from a source the kernel can't (directly) fetch it's going to be no better off than how it is now.

So if storing on the ESP I would say it's better to have systemd-boot and grub parse it and pass it on the kernel command line instead.

**Comment by bluca** [comment]
> But on _modern_ cheap devices? Everyone uses the same 4-5 IBVs. Modern stuff will be on a modern stack.

Yeah I don't think the advice around this has changed. I am pretty sure we do not want to be the first ones to test the theory that it should not be a problem anymore, as we get enough flak as it is.

> So if storing on the ESP I would say it's better to have systemd-boot and grub parse it and pass it on the kernel command line instead.

We are most definitely not going to parse untrusted stuff like this in sd-boot itself. Also appending random stuff that changes to the kernel command line means the measurements change in an unpredicatable way, and having that happen every time the brightness setting is updated is a recipe for breaking tons of stuff.
This is not some critical infrastructure without which nothing works, it's a good-to-have, so I don't see why we need to go out of our way to make it absolutely perfect. Good enough is enough. Having systemd read a credential stored in the ESP is something we can already do, so all that is needed is for userspace to set it, and read it and apply it in the initrd. It doesn't even need any systemd changes, any recent enough release can support all of this already.

**Comment by superm1** [comment]
> Yeah I don't think the advice around this has changed. I am pretty sure we do not want to be the first ones to test the theory that it should not be a problem anymore, as we get enough flak as it is.

What about adding this and making this opt in?

> Also appending random stuff that changes to the kernel command line means the measurements change in an unpredicatable way, and having that happen every time the brightness setting is updated is a recipe for breaking tons of stuff.

Ah, thanks I hadn't considered the implication for that change.
